File: languages/c/queue/linked_queue.c
#include <stdlib.h>

#include "queue.h"

LinkedQueue *get_linked_queue() {
    LinkedQueue *q = (LinkedQueue *) malloc(sizeof(LinkedQueue));
    q->first = NULL;
    q->last = NULL;
    q->size = 0;

    return q;
}

int linked_queue_empty(LinkedQueue *q) {
    return q->size == 0;
}

void linked_queue_enqueue(LinkedQueue *q, int e) {
    if(linked_queue_empty(q)) {
        q->first = (QueueNode *) malloc(sizeof(QueueNode));
        q->first->data = e;
        q->last = q->first;
        q->size++;
    }else{
        QueueNode *s = (QueueNode *) malloc(sizeof(QueueNode));
        s->data = e;
        s->next = NULL;
        q->last->next = s;
        q->last = s;
        q->size++;
    }
}

int linked_queue_dequeue(LinkedQueue *q) {
    if(linked_queue_empty(q)) {
        printf("dequeue error, queue is empty");
        exit(-1);
    }

    int e = q->first->data;
    q->first = q->first->next;
    q->size--;

    return e;
}

int linked_queue_peek(LinkedQueue *q) {
    if(linked_queue_empty(q)) {
        printf("peek error, queue is empty");
        exit(-1);
    }

    return q->first->data;
}

File: src/queue/LinkedQueue.java
/******************************************************************************
 *  Compilation:  javac LinkedQueue.java
 *  Execution:    java LinkedQueue < input.txt
 *  Dependencies: StdIn.java StdOut.java
 *  Data files:   https://algs4.cs.princeton.edu/13stacks/tobe.txt  
 *
 *  A generic queue, implemented using a singly linked list.
 *
 *  % java Queue < tobe.txt 
 *  to be or not to be (2 left on queue)
 *
 ******************************************************************************/

 package queue;

 import java.util.Iterator;
import java.util.NoSuchElementException;

import utils.StdIn;
import utils.StdOut;

 public class LinkedQueue<Item> implements Iterable<Item> {
     private class Node {
         private Item item;
         private Node next;
     }

     private Node first;
     private Node last;
     private int size;

     public LinkedQueue() {
         first = last = null;
         size = 0;

         assert check();
     }

     public boolean isEmpty() {
        //  return size == 0;
         return first == null;
     }

     public int size() {
         return size;
     }

     public Item peek() {
         if(isEmpty()) throw new NoSuchElementException("Queue underflow");
         return first.item;
     }

     public void enqueue(Item item) {
         Node oldlast = last;
         last = new Node();
         last.item = item;
         last.next = null;
         size++;
         if(isEmpty()) first=last;
         else oldlast.next = last;

         assert check();
     }

     public Item dequeue() {
         if(isEmpty()) throw new NoSuchElementException("Queue underflow");
         Item item = first.item;
         first = first.next;
         size--;
         if(isEmpty()) last = null;

         assert check();
         return item;
     }

     public String toString() {
         StringBuilder s = new StringBuilder();
         for(Item item:this) {
             s.append(item + " ");
         }

         return s.toString();
     }

     public Iterator<Item> iterator() {
        return new ListIterator();
     }

     private class ListIterator implements Iterator<Item> {
        private Node current = first;

        public boolean hasNext() {return current != null;}
        public void remove() {throw new UnsupportedOperationException();}

        public Item next() {
            if(!hasNext()) throw new NoSuchElementException();
            Item item = current.item;
            current = current.next;

            return item;
        }
     }

     private boolean check() {
        if(size < 0){
            return false;
        }else if(size == 0) {
            if(first != null) return false;
            if(last != null) return false;
        }else if(size==1) {
            if(first==null || last==null) return false;
            if(first != last) return false;
            if(first.next != null) return false;
        }else{
            if(first==null || last==null) return false;
            if(first == last) return false;

            if(first.next==null) return false;
            if(last.next != null) return false;

            int n = 0;
            for(Node x=first; x!=null; x=x.next)
                n++;
            if(n!=size) return false;

            Node lastNode = first;
            while(lastNode.next != null)
                lastNode = lastNode.next;
            if(lastNode!=last) return false;
        }
        
        return true;
     }

     public static void main(String[] args) {
         LinkedQueue<String> q = new LinkedQueue<String>();
         while(!StdIn.isEmpty()) {
             String item = StdIn.readString();
             if(!item.equals("-")) q.enqueue(item);
             else if(!q.isEmpty())
                StdOut.print(q.dequeue() + " ");
         }

         StdOut.println("(" + q.size() + " left on queue)");
     }
 }